x/mint/keeper/hooks.go

```diff
@@ -0,0 +1,132 @@
+package keeper
+
+import (
+	"context"
+	"fmt"
+
+	"cosmossdk.io/core/event"
+	epochstypes "cosmossdk.io/x/epochs/types"
+	"cosmossdk.io/x/mint/types"
+
+	"github.com/cosmos/cosmos-sdk/telemetry"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// BeforeEpochStart is a hook which is executed before the start of an epoch. It is a no-op for mint module.
+func (k Keeper) BeforeEpochStart(ctx context.Context, epochIdentifier string, epochNumber int64) error {
+	// no-op
+	return nil
+}
+
+// AfterEpochEnd is a hook which is executed after the end of an epoch.
+// This hook should attempt to mint and distribute coins according to
+// the configuration set via parameters. In addition, it handles the logic
+// for reducing minted coins according to the parameters.
+// For an attempt to mint to occur:
+// - given epochIdentifier must be equal to the mint epoch identifier set via parameters.
+// - given epochNumber must be greater than or equal to the mint start epoch set via parameters.
+func (k Keeper) AfterEpochEnd(ctx context.Context, epochIdentifier string, epochNumber int64) error {
+	params, err := k.Params.Get(ctx)
+	if err != nil {
+		return err
+	}
+
+	if epochIdentifier == params.EpochIdentifier {
+		// not distribute rewards if it's not time yet for rewards distribution
+		if epochNumber < params.MintingRewardsDistributionStartEpoch {
+			return nil
+		} else if epochNumber == params.MintingRewardsDistributionStartEpoch {
+			err := k.setLastReductionEpochNum(ctx, epochNumber)
+			if err != nil {
+				return err
+			}
+		}
+		// fetch stored minter & params
+		minter, err := k.Minter.Get(ctx)
+		if err != nil {
+			return err
+		}
+
+		// Check if we have hit an epoch where we update the inflation parameter.
+		// We measure time between reductions in number of epochs.
+		// This avoids issues with measuring in block numbers, as epochs have fixed intervals, with very
+		// low variance at the relevant sizes. As a result, it is safe to store the epoch number
+		// of the last reduction to be later retrieved for comparison.
+		lastReductionEpochNum, err := k.getLastReductionEpochNum(ctx)
+		if err != nil {
+			return err
+		}
+		if epochNumber >= params.ReductionPeriodInEpochs+lastReductionEpochNum {
+			// Reduce the reward per reduction period
+			minter.EpochProvisions = minter.NextEpochProvisions(params)
+			err := k.Minter.Set(ctx, minter)
+			if err != nil {
+				return err
+			}
+			err = k.setLastReductionEpochNum(ctx, epochNumber)
+			if err != nil {
+				return err
+			}
+		}
+
+		// mint coins, update supply
+		mintedCoin := minter.EpochProvision(params)
```

Comment:
how is this calculated? it should still be a function of how much is staked no? this change might be too fundamental for chains to adopt. Also it changes an economic design. We should reconsider this change and possibly revert it